### ISSUE TITLE
CommonJS behavior for `require`

### ISSUE THREAD

**Issue by mafredri** [kasper/phoenix#91]
Would it be possible to implement [CommonJS](http://requirejs.org/docs/commonjs.html) behavior for `require`?

I'm currently writing my Phoenix configuration in TypeScript (also created [typings](https://github.com/mafredri/phoenix-typings) for it) and possible module targets for compilation are CommonJS, AMD, SystemJS, UMD or ES2015. Since most of these are meant for the browser, and ES2015 isn't supported in Phoenix, CommonJS is the most suitable pick.

The problem is, however, that CommonJS relies on `require`s being modules (using `exports = {}` syntax) and uses module resolution as well (e.g. `require('./window')` would import `window.js` from the same directory).

I'm currently using the following workaround at the top of my phoenix configuration:

```js
(function (global, requirePhoenix) {
  global.require = function (mod) {
    global.exports = {};
    requirePhoenix(mod + '.js');
    return global.exports;
  };
})(this, require);
```

This works, but it's far from optimal, it expects a certain format and is unable to perform real module resolution.

I realize changing `require()` behavior would break backwards compat in Phoenix, but it might be possible to extend it and introduce an opt-in for CommonJS behavior. For example, by default require works as before, if no file is found, perform module resolution. If a module is detected (e.g. uses exports) then import as a CommonJS module.

One benefit from this is that it could make it much easier to utilize npm modules in Phoenix.

Either way, big thanks for Phoenix, been enjoying it so-far 👍 !

**Comment by jasonm23** [comment]
It's much better if you simply make or grunt or gulp your config down to JS.

The hash bang provision is intended as a convenient way to get preprocessing.

Explicit support everyone's pet module system / or preprocessing language isn't a good use of development time, because it's easier for users to tool their source, and get their desired config, and also keep up with whatever latest greatest JS flavor they happen to be into.

**Comment by jasonm23** [comment]
Ultimately it's up to @kasper, but I wanted to address this, because it's better to leave this sort of thing for users to deal with.

It's much better for Phoneix development to build features which are opening up more of the machine/OS X.

**Comment by mafredri** [comment]
I appreciate the honesty. It's not that I'm avoiding tooling, I already compile my TypeScript source down to JS. The fact is, though, that Phoenix doesn't support any standard form of module loading (not that it has to), and the name require can be confusing as it does not do what users would expect.

Without this feature I have to rely on hackish solutions because I'm using TypeScript. But I'm fine with whatever kasper decides, though. I thought this could make a nice addition, but there's no point in implementing something there isn't enough need for (as you said).

**Comment by jasonm23** [comment]
Thanks for taking my note in the intended spirit.

The fragility of JS tool chains (ie. their propensity to churn) is a bit of a rabbit hole.

**Comment by kasper** [comment]
@mafredri Thanks for the kind words and I especially like your work on providing typings for Phoenix! Feel free to add a link to the Wiki. :)

Your points are valid and I’ve occasionally thought about this. You’re right that in this context require isn’t modular as with Node.js or CommonJS. It’s simply just a convenience import so you can split your code — relatable to other languages. Some might expect otherwise. Whether it should support modular exporting is up for discussion. Your solution is short and simple, although I understand it might feel more like a workaround.

I tend to agree with @jasonm23. Likely this would only benefit a small portion of the users. Those users can extend their setup with tools such as Grunt or Gulp, using language compilers as preprocessors or even writing their own preprocessors.

I would also prefer to keep the core as simple and efficient as possible. However, what I would like to concentrate next is creating a JavaScript-library to extend the core with reusable and commonly used features. (Maybe even creating a default configuration.) 

Regarding this, I have thought about adding a very minimal way to require libraries/configurations/files directly from GitHub-repositories. Maybe this could be implemented as a library… Then again, I don’t want to complicate things too much. This is a very borderline case.

Every developer has their own tools of trade. Obviously it doesn’t make sense to support everything time wise or technically, but I also want to make sure that people can get the most out of this cool project.

Have you found some npm-packages to be especially valuable with Phoenix?

**Comment by mafredri** [comment]
Thanks for the thorough writeup @kasper. I appreciate your thoughts on the matter and can relate. I'm definitely looking forward to future development of Phoenix, even if this is not part of it :)! I like the library idea and do have some ideas for new features that could be useful. I'll try to open up some issues for discussion when I have a bit more time to properly define them.

As it turns out, adding more modules caused problems with my intial solution (the `export` that the modules refer to gets replaced since they are not wrapped in IIFE's). I created a [fake AMD-style loader](https://gist.github.com/mafredri/be2c6b10956a28c7b1d66d50808e75eb) that allows me to compile my TypeScript source into a single file with AMD module loading.

While writing the AMD loader above, I had the idea to check for the presence of files by putting the require in a `try/catch` statement, but it doesn't seem possible to filter out the error messages. It would be nice to be able to catch this error, or detect files. But I'm currently satisfied with this setup, so I'd put this on the nice-to-have list.

> Feel free to add a link to the Wiki. :)

Thanks, will do 👍 !

> Have you found some npm-packages to be especially valuable with Phoenix?

That is in fact a very good question, can't think of anything specific. Even then it'd probably just be some handy utility library like Underscore, heh.

**Comment by kasper** [comment]
Discussion and throwing ideas around is always good! 

It could be rather useful if `require` would actually throw an error for instance when the file cannot be resolved. This would extend the possibilities for similar use cases. I’ll add this to the TODOs. Thanks!

**Comment by kasper** [comment]
@mafredri I’ve been thinking about this. I wonder if there would be an easy way to enable you to accomplish this yourself in the JavaScript scope without it feeling like a workaround. 

Basically, if `require` would return the “module”, you could implement the rest yourself. Loading a script in JavaScriptCore will return the last evaluated value. Maybe, it could be possibly to use this value to return the “module”. Maybe a wrapper function or something that would encapsulate the scope instead of polluting the global scope. This also would keep the existing behaviour as is.

**Comment by mafredri** [comment]
@kasper interesting idea. I'm not quite sure what you mean with returning the last evaluated value, but I can't see how this doesn't change existing behavior? If `require` loads scripts in their own scope, then those functions defined by the required script would not be accessible after a require, as I believe they currently are?

Here's a very simplistic example of a CommonJS module structure:

```js
// a.js
function loga() { Phoenix.log('hello from a'); }
exports.loga = function() { loga(); }

// b.js
function logb() { Phoenix.log('hello from b'); }
exports = function() { logb(); }

// phoenix.js
var a = require('./a.js');
var b = require('./b.js');

a.loga(); // hello from a
b();      // hello from b
```

I'm not quite clear as to how this change would help in this particular case. For example, `exports` is never defined and if it was defined in `phoenix.js`, the exports object / function would probably not retain a reference to the original functions (`loga` / `logb`). Here's a simplistic illustration of how `phoenix.js` would change in this case:

```js
// phoenix.js
var exports = {};
var a = require('./a.js');
a = exports;
exports = {};
var b = require('./b.js');
b = exports;

a.loga(); // loga not found
b();      // logb not found
```

I'm open to work on this further though, if you think we can solve it neatly :D!

**Comment by kasper** [comment]
@mafredri Basically, I don’t really want to implement a modular require that would need to keep track of the modules within the core or neither do I really want to break the existing simple behaviour. 

However, I image you could very well override this behaviour yourself in the JavaScript scope (as you partly have done) and implement “true” modularity and keep the states alive as a library with only a few changes to the core.

When JavaScriptCore loads a script, it will return the last evaluated value. Currently we do nothing with it, but `require` could very well return it. This wouldn’t break anything. I haven’t tested this yet — I wonder how this behaviour actually works, but given the following example, I would assume the last evaluated value to be the value of `b`, so 2.

```javascript
var a = 1;
var b = 2;
```

Obviously, the problem is that this can’t be used with the usual “exports”-strategy. It would pollute the global scope and it wouldn’t really be useful to require the exports-statement to be the last value in the file. 

A.) But, I wonder if something like this for would work.

```javascript
/* something.js */

module(function () {
  return {};
});
```

If in this case the last evaluated value would indeed be the value of `module` (which would just be a wrapper that would return the value of the function), you could use this to create a new scope. This wouldn’t change any existing behaviour, you could use the return value if you want and extend the behaviour to keep track of the states of the modules. Though, I’m not sure if this would work in practise and obviously this would not be a standard solution. Just an interesting thought that came into my mind.

B.) To support the “exports”-strategy, we would need to create a new context for each file we load, but I imagine this wouldn’t need to break existing behaviour either.

1. Create a new context for each `require`.
2. Load and evaluate the file to this new scope.
3. Check if “exports” has been set.
4. If it has been set, return the value of it through `require`.
5. If it hasn’t been set, just load and evaluate the file to the global scope.
6. The new context would be then destroyed.

This still would not keep track of the modules or their state, but you would be able again to extend this yourself. Even resolving the modules without the full path.

**Comment by mafredri** [comment]
> Basically, I don’t really want to implement a modular require that would need to keep track of the modules within the core or neither do I really want to break the existing simple behaviour.

I figured as much :)!

> A.) But, I wonder if something like this for would work.

Hmm, I think this could work very well with any of `AMD`, `UMD` or `SystemJS`. Whereas not so much for `CommonJS` since the declarations aren't wrapped.

> B.) To support the “exports”-strategy, we would need to create a new context for each file we load, but I imagine this wouldn’t need to break existing behaviour either.

This on the other hand sounds like just what is needed to support `CommonJS`, nice! Would `require` take an extra parameter (`scope`) in this case? Or maybe just a `require.call(scope, 'path');`?

**Comment by kasper** [comment]
@mafredri Yes, I think the latter is a better choice. No, the scope would be created in runtime. Basically this shouldn’t break anything. `Require` would just return the value of `exports` and ensure the file is evaluated in a new scope — if this is the strategy.
